sklearn/metrics/pairwise.py

```diff
@@ -256,6 +264,166 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    >>> nan = float("NaN")
+    >>> X = [[0, 1], [1, nan]]
+    >>> # distance between rows of X
+    >>> masked_euclidean_distances(X, X)
+    array([[ 0.        ,  1.41421356],
+           [ 1.41421356,  0.        ]])
+
+    >>> # get distance to origin
+    >>> masked_euclidean_distances(X, [[0, 0]])
+    array([[ 1.        ],
+           [ 1.41421356]])
+
+    See also
+    --------
+    paired_distances : distances betweens pairs of elements of X and Y.
+    """
+    # Check and except sparse matrices
+    if issparse(X) or (Y is not None and issparse(Y)):
+        raise ValueError(
+            "Missing value support for sparse matrices not added yet")
+
+    # NOTE: force_all_finite=False allows not only NaN but also inf/-inf
+    X, Y = check_pairwise_arrays(X, Y,
+                                 force_all_finite=False, copy=copy)
+    if (np.any(np.isinf(X.data)) or
+            (Y is not None and np.any(np.isinf(Y.data)))):
+        raise ValueError(
+            "+/- Infinite values are not allowed.")
+
+    # Check if any rows have only missing value
+    if np.any(_get_mask(X, missing_values).sum(axis=1) == X.shape[1])\
+            or (Y is not None and np.any(_get_mask(Y, missing_values).sum(
+                axis=1) == Y.shape[1])):
+        raise ValueError("One or more rows only contain missing values.")
+    #
+    # if kill_missing:
+    #     if X_norm_squared is not None:
+    #         XX = check_array(X_norm_squared)
+    #         if XX.shape == (1, X.shape[0]):
+    #             XX = XX.T
+    #         elif XX.shape != (X.shape[0], 1):
+    #             raise ValueError(
+    #                 "Incompatible dimensions for X and X_norm_squared")
+    #     else:
+    #         XX = row_norms(X, squared=True)[:, np.newaxis]
+    #
+    #     if X is Y:  # shortcut in the common case euclidean_distances(X, X)
+    #         YY = XX.T
+    #     elif Y_norm_squared is not None:
+    #         YY = np.atleast_2d(Y_norm_squared)
+    #
+    #         if YY.shape != (1, Y.shape[0]):
+    #             raise ValueError(
+    #                 "Incompatible dimensions for Y and Y_norm_squared")
+    #     else:
+    #         YY = row_norms(Y, squared=True)[np.newaxis, :]
+    #
+    #     distances = safe_sparse_dot(X, Y.T, dense_output=True)
+    #     distances *= -2
+    #     distances += XX
+    #     distances += YY
+    #     np.maximum(distances, 0, out=distances)
+
+    # else:
+    if missing_values != "NaN" and \
+            (np.any(_get_mask(X.data, "NaN")) or
+                np.any(_get_mask(Y.data, "NaN"))):
+        raise ValueError(
+            "NaN values present but missing_value = {0}".format(
+                missing_values))
+
+    # Get missing mask for X
+    mask_X = _get_mask(X, missing_values)
+
+    # Get Y.T mask and anti-mask and set Y.T's missing to zero
+    YT = Y.T
+    mask_YT = _get_mask(YT, missing_values)
+    NYT = (~mask_YT).astype(np.int8)
```

Comment:
No I did that because leaving it as bool was returning incorrect values. I think the issue is that dot product of bool matrices returns a bool which of course means that we do not get the sum of True as a sum of ones, as we would want for individual dot products.